src/transforms/projectionpursuit.jl

```diff
@@ -0,0 +1,165 @@
+  return I
+end
+
+function optimaxis(transform::ProjectionPursuit,table)
+  _, q = size(table)
+  Iₘₐₓ, iₘₐₓ = 0, 0
+  for i = 1:q
+    eᵢ = [1. * (j==i) for j=1:q]
+    I = pindex(transform, table, eᵢ)
+    if I > Iₘₐₓ
+      Iₘₐₓ, iₘₐₓ = I, i
+    end
+  end
+  return [1. *(iₘₐₓ==i) for i = 1:q]
+end
+
+function optimdiag(transform::ProjectionPursuit,table, ᾱ)
+  l = length(ᾱ)
+  I₀ = pindex(transform, table, ᾱ)
+  for i in 1:l
+    eᵢ = [1. * (j==i) for j=1:l]
+    f₊ = pindex(transform, table, (1/√(2+2ᾱ[i]))*(ᾱ + eᵢ))
+    f₋ = pindex(transform, table, (1/√(2-2ᾱ[i]))*(ᾱ - eᵢ)) * 
+      (ᾱ[i] != 1)
+    f = (f₊>f₋)*f₊ + (f₊<=f₋)*f₋
+    s = (f₊>f₋)*1 + (f₊<=f₋)*(-1)
+    if f > pindex(transform, table, ᾱ)
+      ᾱ = (1/√(2+2s*ᾱ[i]))*(ᾱ + s*eᵢ)
+    end
+  end
+  if pindex(transform, table,ᾱ) >= I₀
+    return ᾱ
+  else
+    return optimaxis(transform, table)
+  end
+end
+
+function optimgen(transform::ProjectionPursuit, table, α₀)
+  f(α) = -pindex(transform, table, α)
+  op = optimize(f,α₀)
+  Optim.minimizer(op)
+end
+
+function optim(transform::ProjectionPursuit, table)
+  α_axis = optimaxis(transform, table)
+  α_diag = optimdiag(transform, table, α_axis)
+  optimgen(transform, table, α_diag)  
+end
+
+#Transforms the cdf of a convex combination of r.v. to a standard normal cdf.
+function structure_removal(transform::ProjectionPursuit, table, α)
+  l = length(α)
+  U = qr(hcat(α,rand(l,l-1)))
+  if norm(diag(U.R)) < transform.tol
+    return print("Error: the random matrix is singular. Try again")
+  end
+  T = Tables.matrix(table) * U.Q
+  df = DataFrame(T,:auto)
+  T[:,1] = apply(Quantile(),df)[1][:,1]
+  return T*U.Q'
+end
+
+#Gives the array that will be used to stop the loop of structure 
+#removal by computing the projection index on k samples of multinomial 
+#distribution (size of sample kxn) along q random directions.
+function gauss_pmatrix(transform::ProjectionPursuit,N,q)
+  I = rand(transform.n,q)
+  d = MvNormal(zeros(q), Diagonal(ones(q)))
+  Q = qr(rand(q,q)).Q
+  for i = 1:transform.n
+    S = DataFrame(Matrix(rand(d,N)'),:auto)
+    for j = 1:q
+      I[i,j] = pindex(transform, S, Q[:,j])
+    end
+  end
+  I, Q
+end
+
+#Computes the projection index of each loop dataset (with possibly 
+#modified structure) along q random directions and compares with the 
+#array given by gaussian_projection_matrix.
+function stop_criteria(transform::ProjectionPursuit,matrix,I,Q)
+  k, q = size(I)
+  switch = true
+  for i = 1:q
+    b = pindex(transform,DataFrame(matrix,:auto),Q[:,i])
+    switch *= quantile(I[1:k-1,i],transform.perc) > b
+  end
+  return !switch
+end
+
+#Generates the loop for the structure removal process
+function apply(transform::ProjectionPursuit, table) #ppmt(D,pi,J,k)
+  seq = Quantile() → EigenAnalysis(:VDV)
+  Z = Matrix(apply(seq, table)[1])
+  N,q = size(table)
+  I, Q = gauss_pmatrix(transform, N, q) 
+  count = 0
+  while stop_criteria(transform, Z, I, Q)
+    α = optim(transform, table)
```

Comment:
I think auxiliary functions should operate on `M = Tables.matrix(table)` instead of the table itself.
This prevents the table from being converted into matrix several times.